File: ios/stereoCamera/stereoCamera/Comm/Commands/Command.swift
//
//  Command.swift
//  stereoCamera
//
//  Created by hallmarklabs on 6/15/18.
//  Copyright © 2018 cody. All rights reserved.
//

import Foundation

class Command
{
    static var nextId:Int = 1
    
    public var cmdtype:CommandTypes = CommandTypes.NONE
    public var id:Int
    public var isResponse:Bool = false
    
    var expectsResponse:Bool = false
    
    init()
    {
        id = Command.nextId
        Command.nextId = Command.nextId + 1
    }
    
    func send(comm:Comm) -> Bool
    {    
        var bytes:[UInt8] = []
        bytes += Bytes.toByteArray(cmdtype)
        bytes += Bytes.toByteArray(id)
        bytes += Bytes.toByteArray(isResponse)
        
        let sz = comm.write(buf: bytes)
        if (sz <= 0)
            { return false }
        
        return true
    }
    
    func receive(comm:Comm) -> Bool
    {
        let (buf, sz) = comm.read(sz: 9)
        
        if (sz <= 0)
            { return false }
        
        id = Bytes.fromByteArray(buf)
        isResponse = Bytes.fromByteArray(buf, 8)
        
        return true
    }
    
    
    
    func onResponse(command:Command)
    {}
}

class DefaultCommand : Command
{
    override init()
    {
        super.init()
        self.cmdtype = CommandTypes.NONE
    }
}

class CommandFactory
{
    static func build(type: CommandTypes) -> Command
    {
        switch type
        {
        case CommandTypes.PING:
            return Ping()
        case CommandTypes.HANDSHAKE:
            return Handshake()
        case CommandTypes.SEND_VERSION:
            return Version()
        case CommandTypes.RECEIVE_STATUS:
            return SendStatus()
        case CommandTypes.SET_FACING:
            return SetFacing()
        case CommandTypes.SET_OVERLAY:
            return SetOverlay()
        case CommandTypes.SET_ZOOM:
            return SetZoom()
        case CommandTypes.RECEIVE_ZOOM:
            return SendZoom()
        case CommandTypes.FIRE_SHUTTER:
            return FireShutter()
        case CommandTypes.SEND_PROCESSED_PHOTO:
            return SendPhoto()
        case CommandTypes.DISCONNECT:
            return Disconnect()
        case CommandTypes.RECEIVE_DISCONNECT:
            return SendDisconnect()
        case CommandTypes.CONNECTION_PAUSE:
            return ConnectionPause()
        case CommandTypes.RECEIVE_CONNECTION_PAUSE:
            return SendConnectionPause()
        case CommandTypes.RECEIVE_GRAVITY:
            return SendGravity()
        case CommandTypes.SET_CAPTURE_QUALITY:
            return SetCaptureQuality()
        case CommandTypes.LATENCY_CHECK:
            return LatencyTest()
        case CommandTypes.ID:
            return ID()
        default:
            print("CommandFactory unable to construct command of type " + type.description)
            return DefaultCommand()
        }
    }
}

File: android/app/src/main/java/com/munger/stereocamera/ip/command/Command.java
package com.munger.stereocamera.ip.command;

import com.munger.stereocamera.ip.SocketCtrl;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;

public class Command
{
	protected static int nextId = 1;

	public enum Type
	{
		NONE,
		FIRE_SHUTTER,
		PING,
		LATENCY_CHECK,
		RECEIVE_ANGLE_OF_VIEW,
		SET_ZOOM,
		SET_FACING,
		SET_OVERLAY,
		SEND_VERSION,
		RECEIVE_ZOOM,
		RECEIVE_STATUS,
		RECEIVE_GRAVITY,
		RECEIVE_ORIENTATION,
		CONNECTION_PAUSE,
		DISCONNECT,
		SEND_PROCESSED_PHOTO,
		RECEIVE_PREVIEW_FRAME,
		HANDSHAKE,
		SET_CAPTURE_QUALITY,
		ID
	}

	public Type cmdtype = Type.NONE;
	public int id;
	public boolean isResponse = false;

	protected boolean expectsResponse = false;

	public Command()
	{
		id = Command.nextId;
		Command.nextId++;
	}

	public boolean send(Comm comm)
	{
		try
		{
			comm.putByte((byte) cmdtype.ordinal());
			comm.putLong(id);
			comm.putByte((byte) ((isResponse) ? 1 : 0));
		}
		catch (IOException e){
			return false;
		}

		return true;
	}

	public boolean receive(Comm comm)
	{
		try
		{
			id = (int) comm.getLong();
			isResponse = comm.getBoolean();
		}
		catch(IOException e){
			return false;
		}

		return true;
	}

	public void onResponse(Command command)
	{}
}
